# Projeto de Implementação 1

**Aluno:** Igor Carvalho da Silva  
**Matrícula:** 21951288

## Parte 1

### 1 - Instalando Hadoop

In [1]:
!wget https://downloads.apache.org/hadoop/common/hadoop-3.3.4/hadoop-3.3.4.tar.gz

--2022-11-14 21:16:10--  https://downloads.apache.org/hadoop/common/hadoop-3.3.4/hadoop-3.3.4.tar.gz
Resolving downloads.apache.org (downloads.apache.org)... 88.99.95.219, 135.181.214.104, 2a01:4f8:10a:201a::2, ...
Connecting to downloads.apache.org (downloads.apache.org)|88.99.95.219|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 695457782 (663M) [application/x-gzip]
Saving to: ‘hadoop-3.3.4.tar.gz’

hadoop-3.3.4.tar.gz 100%[===================>] 663.24M  25.0MB/s    in 27s     

2022-11-14 21:16:37 (24.6 MB/s) - ‘hadoop-3.3.4.tar.gz’ saved [695457782/695457782]



In [2]:
!tar -xzvf hadoop-3.3.4.tar.gz

A saída de streaming foi truncada nas últimas 5000 linhas.
hadoop-3.3.4/share/doc/hadoop/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/apidocs/org/apache/hadoop/yarn/server/nodemanager/containermanager/resourceplugin/deviceframework/package-tree.html
hadoop-3.3.4/share/doc/hadoop/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/apidocs/org/apache/hadoop/yarn/server/nodemanager/containermanager/resourceplugin/deviceframework/DeviceMappingManager.html
hadoop-3.3.4/share/doc/hadoop/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/apidocs/org/apache/hadoop/yarn/server/nodemanager/containermanager/resourceplugin/deviceframework/DeviceResourceHandlerImpl.html
hadoop-3.3.4/share/doc/hadoop/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/apidocs/org/apache/hadoop/yarn/server/nodemanager/containermanager/resourceplugin/deviceframework/ShellWrapper.html
hadoop-3.3.4/share/doc/hadoop/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemana

In [3]:
#copy  hadoop file to user/local
!cp -r hadoop-3.3.4/ /usr/local/

### 2 - Configurando Hadoop’s Java Home

In [4]:
#To find the default Java path
!readlink -f /usr/bin/java | sed "s:bin/java::"

/usr/lib/jvm/java-11-openjdk-amd64/


In [5]:
!echo 'export JAVA_HOME=/usr/lib/jvm/java-11-openjdk-amd64/' >> /usr/local/hadoop-3.3.4/etc/hadoop/hadoop-env.sh

### 3 - Rodando Hadoop

In [6]:
!/usr/local/hadoop-3.3.4/bin/hadoop

Usage: hadoop [OPTIONS] SUBCOMMAND [SUBCOMMAND OPTIONS]
 or    hadoop [OPTIONS] CLASSNAME [CLASSNAME OPTIONS]
  where CLASSNAME is a user-provided Java class

  OPTIONS is none or any of:

buildpaths                       attempt to add class files from build tree
--config dir                     Hadoop config directory
--debug                          turn on shell script debug mode
--help                           usage information
hostnames list[,of,host,names]   hosts to use in slave mode
hosts filename                   list of hosts to use in slave mode
loglevel level                   set the log4j level for this command
workers                          turn on worker mode

  SUBCOMMAND is one of:


    Admin Commands:

daemonlog     get/set the log level for each daemon

    Client Commands:

archive       create a Hadoop archive
checknative   check native Hadoop and compression libraries availability
classpath     prints the class path needed to get the Hadoop jar and the
    

In [7]:
!mkdir ~/input
!cp /usr/local/hadoop-3.3.4/etc/hadoop/*.xml ~/input

In [8]:
!ls ~/input

capacity-scheduler.xml	hdfs-rbf-site.xml  kms-acls.xml     yarn-site.xml
core-site.xml		hdfs-site.xml	   kms-site.xml
hadoop-policy.xml	httpfs-site.xml    mapred-site.xml


In [9]:
!/usr/local/hadoop-3.3.4/bin/hadoop jar /usr/local/hadoop-3.3.4/share/hadoop/mapreduce/hadoop-mapreduce-examples-3.3.4.jar grep ~/input ~/grep_example 'allowed[.]*'

2022-11-14 21:23:10,918 INFO impl.MetricsConfig: Loaded properties from hadoop-metrics2.properties
2022-11-14 21:23:11,199 INFO impl.MetricsSystemImpl: Scheduled Metric snapshot period at 10 second(s).
2022-11-14 21:23:11,200 INFO impl.MetricsSystemImpl: JobTracker metrics system started
2022-11-14 21:23:11,490 INFO input.FileInputFormat: Total input files to process : 10
2022-11-14 21:23:11,541 INFO mapreduce.JobSubmitter: number of splits:10
2022-11-14 21:23:11,998 INFO mapreduce.JobSubmitter: Submitting tokens for job: job_local984878570_0001
2022-11-14 21:23:11,998 INFO mapreduce.JobSubmitter: Executing with tokens: []
2022-11-14 21:23:12,310 INFO mapreduce.Job: The url to track the job: http://localhost:8080/
2022-11-14 21:23:12,311 INFO mapreduce.Job: Running job: job_local984878570_0001
2022-11-14 21:23:12,322 INFO mapred.LocalJobRunner: OutputCommitter set in config null
2022-11-14 21:23:12,335 INFO output.FileOutputCommitter: File Output Committer Algorithm version is 2
2022-1

In [10]:
!cat ~/grep_example/*

22	allowed.
1	allowed


Download 20newsgroups dataset available at http://qwone.com/~jason/20Newsgroups.

In [11]:
!wget http://qwone.com/~jason/20Newsgroups/20news-18828.tar.gz
!tar -xzvf 20news-18828.tar.gz

A saída de streaming foi truncada nas últimas 5000 linhas.
20news-18828/sci.space/61079
20news-18828/sci.space/60809
20news-18828/sci.space/61497
20news-18828/sci.space/60156
20news-18828/sci.space/61475
20news-18828/sci.space/59873
20news-18828/sci.space/61541
20news-18828/sci.space/61489
20news-18828/sci.space/61133
20news-18828/sci.space/62418
20news-18828/sci.space/61289
20news-18828/sci.space/60864
20news-18828/sci.space/61443
20news-18828/sci.space/60189
20news-18828/sci.space/60251
20news-18828/sci.space/61356
20news-18828/sci.space/61255
20news-18828/sci.space/61345
20news-18828/sci.space/61382
20news-18828/sci.space/60847
20news-18828/sci.space/62481
20news-18828/sci.space/60783
20news-18828/sci.space/60854
20news-18828/sci.space/62131
20news-18828/sci.space/60921
20news-18828/sci.space/62317
20news-18828/sci.space/61372
20news-18828/sci.space/61406
20news-18828/sci.space/60247
20news-18828/sci.space/62114
20news-18828/sci.space/60791
20news-18828/sci.space/61362
20news-18828/

#### Hadoop Streaming

In [12]:
!find / -name 'hadoop-streaming*.jar'

/usr/local/hadoop-3.3.4/share/hadoop/tools/lib/hadoop-streaming-3.3.4.jar
/usr/local/hadoop-3.3.4/share/hadoop/tools/sources/hadoop-streaming-3.3.4-test-sources.jar
/usr/local/hadoop-3.3.4/share/hadoop/tools/sources/hadoop-streaming-3.3.4-sources.jar
find: ‘/proc/46/task/46/net’: Invalid argument
find: ‘/proc/46/net’: Invalid argument
/content/hadoop-3.3.4/share/hadoop/tools/lib/hadoop-streaming-3.3.4.jar
/content/hadoop-3.3.4/share/hadoop/tools/sources/hadoop-streaming-3.3.4-test-sources.jar
/content/hadoop-3.3.4/share/hadoop/tools/sources/hadoop-streaming-3.3.4-sources.jar


In [13]:
!wget https://github.com/anjalysam/Hadoop/raw/master/mapper.py
!wget https://github.com/anjalysam/Hadoop/raw/master/reducer.py

--2022-11-14 21:23:44--  https://github.com/anjalysam/Hadoop/raw/master/mapper.py
Resolving github.com (github.com)... 140.82.112.3
Connecting to github.com (github.com)|140.82.112.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/anjalysam/Hadoop/master/mapper.py [following]
--2022-11-14 21:23:45--  https://raw.githubusercontent.com/anjalysam/Hadoop/master/mapper.py
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.111.133, 185.199.109.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 767 [text/plain]
Saving to: ‘mapper.py’

mapper.py           100%[===================>]     767  --.-KB/s    in 0s      

2022-11-14 21:23:45 (39.7 MB/s) - ‘mapper.py’ saved [767/767]

--2022-11-14 21:23:45--  https://github.com/anjalysam/Hadoop/raw/master/reducer.py
Resolving github.

In [14]:
!chmod u+rwx /content/mapper.py
!chmod u+rwx /content/reducer.py

In [15]:
!/usr/local/hadoop-3.3.4/bin/hadoop jar /usr/local/hadoop-3.3.4/share/hadoop/tools/lib/hadoop-streaming-3.3.4.jar -input /content/20news-18828/alt.atheism/49960 -output /content/output/20news -file /content/mapper.py  -file /content/reducer.py  -mapper 'python mapper.py'  -reducer 'python reducer.py'

2022-11-14 21:23:54,817 WARN streaming.StreamJob: -file option is deprecated, please use generic option -files instead.
packageJobJar: [/content/mapper.py, /content/reducer.py] [] /tmp/streamjob7305222058422311782.jar tmpDir=null
2022-11-14 21:23:55,595 INFO impl.MetricsConfig: Loaded properties from hadoop-metrics2.properties
2022-11-14 21:23:55,807 INFO impl.MetricsSystemImpl: Scheduled Metric snapshot period at 10 second(s).
2022-11-14 21:23:55,807 INFO impl.MetricsSystemImpl: JobTracker metrics system started
2022-11-14 21:23:55,827 WARN impl.MetricsSystemImpl: JobTracker metrics system already initialized!
2022-11-14 21:23:56,042 INFO mapred.FileInputFormat: Total input files to process : 1
2022-11-14 21:23:56,069 INFO mapreduce.JobSubmitter: number of splits:1
2022-11-14 21:23:56,370 INFO mapreduce.JobSubmitter: Submitting tokens for job: job_local790952758_0001
2022-11-14 21:23:56,370 INFO mapreduce.JobSubmitter: Executing with tokens: []
2022-11-14 21:23:56,773 INFO mapred.Loca

In [16]:
!ls /content/output/20news

part-00000  _SUCCESS


In [17]:
!cat /content/output/20news/part-00000

034529887x	1
0511211216	1
071	5
080182494x	1
0801834074	1
0877226423	1
0877227675	1
0908	1
0910309264	1
1	1
10	1
11	1
1266	1
1271	1
14	1
140195	1
14215	1
14226	1
142282197	1
17701900	1
1881	1
1977	1
1981	1
1986	1
1988	1
1989	1
1990	1
1992	1
20th	1
226	1
24hour	1
2568900	1
272	1
273	1
2nd	1
3005	1
316	1
372	1
3d	1
3nl	1
4	1
41	2
430	2
4581244	1
4679525	1
490	1
495	2
4rh	1
4rl	1
512	2
53701	1
541	1
59	1
608	1
664	1
700	1
702	1
7119	1
716	1
7215	1
7251	1
750	1
7723	1
787140195	1
787522973	1
831	1
8372475	1
88	1
880	2
8964079	1
8ew	1
91605	1
aah	1
aap	2
abortions	1
absurdities	1
accompanied	1
accounts	1
address	1
addresses	2
adulteries	1
aesthetics	1
african	3
africanamericans	1
agnostic	1
al	1
alien	1
allen	2
also	3
altatheism	1
altatheismarchivename	1
altatheismmoderated	1
alternate	1
alternative	1
although	2
america	1
american	5
americans	2
amherst	1
amongst	1
amusing	1
ancient	1
andor	1
another	1
anselm	1
anthology	3
anyone	1
appendix	2
approachable	1
archive	1
archivename	1
archives	1

## Parte 2

### Baixando coleções

In [18]:
!wget -P collections/ https://raw.githubusercontent.com/igroks/bd2/main/projeto-de-impletacao-1/shakespeare.txt
!wget -P collections/ https://snap.stanford.edu/data/p2p-Gnutella08.txt.gz

--2022-11-14 21:24:13--  https://raw.githubusercontent.com/igroks/bd2/main/projeto-de-impletacao-1/shakespeare.txt
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 122780 (120K) [text/plain]
Saving to: ‘collections/shakespeare.txt’

shakespeare.txt     100%[===================>] 119.90K  --.-KB/s    in 0.02s   

2022-11-14 21:24:14 (5.76 MB/s) - ‘collections/shakespeare.txt’ saved [122780/122780]

--2022-11-14 21:24:14--  https://snap.stanford.edu/data/p2p-Gnutella08.txt.gz
Resolving snap.stanford.edu (snap.stanford.edu)... 171.64.75.80
Connecting to snap.stanford.edu (snap.stanford.edu)|171.64.75.80|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 63491 (62K) [application/x-gzip]
Saving to: ‘collections/p2p-Gnutella08.txt.gz’

p2p-Gn

In [19]:
!gunzip collections/p2p-Gnutella08.txt.gz

### Baixando o mapper e reducer (v2)

In [20]:
!wget https://raw.githubusercontent.com/igroks/bd2/main/projeto-de-impletacao-1/mapper-v2.py
!wget https://raw.githubusercontent.com/igroks/bd2/main/projeto-de-impletacao-1/reducer-v2.py

--2022-11-14 21:24:22--  https://raw.githubusercontent.com/igroks/bd2/main/projeto-de-impletacao-1/mapper-v2.py
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 344 [text/plain]
Saving to: ‘mapper-v2.py’

mapper-v2.py        100%[===================>]     344  --.-KB/s    in 0s      

2022-11-14 21:24:22 (13.5 MB/s) - ‘mapper-v2.py’ saved [344/344]

--2022-11-14 21:24:23--  https://raw.githubusercontent.com/igroks/bd2/main/projeto-de-impletacao-1/reducer-v2.py
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.110.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 616 [text/plain]
Saving 

In [21]:
!chmod u+rwx /content/mapper-v2.py
!chmod u+rwx /content/reducer-v2.py

### Questão 1
Qual o número de palavras distintas na coleção?

Shakespeare:

In [22]:
!/usr/local/hadoop-3.3.4/bin/hadoop jar /usr/local/hadoop-3.3.4/share/hadoop/tools/lib/hadoop-streaming-3.3.4.jar -input /content/collections/shakespeare.txt -output /content/output/shakespeare -file /content/mapper-v2.py  -file /content/reducer-v2.py  -mapper 'python mapper-v2.py'  -reducer 'python reducer-v2.py'

2022-11-14 21:24:32,930 WARN streaming.StreamJob: -file option is deprecated, please use generic option -files instead.
packageJobJar: [/content/mapper-v2.py, /content/reducer-v2.py] [] /tmp/streamjob5802753789832190049.jar tmpDir=null
2022-11-14 21:24:33,645 INFO impl.MetricsConfig: Loaded properties from hadoop-metrics2.properties
2022-11-14 21:24:33,858 INFO impl.MetricsSystemImpl: Scheduled Metric snapshot period at 10 second(s).
2022-11-14 21:24:33,858 INFO impl.MetricsSystemImpl: JobTracker metrics system started
2022-11-14 21:24:33,884 WARN impl.MetricsSystemImpl: JobTracker metrics system already initialized!
2022-11-14 21:24:34,040 INFO mapred.FileInputFormat: Total input files to process : 1
2022-11-14 21:24:34,062 INFO mapreduce.JobSubmitter: number of splits:1
2022-11-14 21:24:34,380 INFO mapreduce.JobSubmitter: Submitting tokens for job: job_local1635643757_0001
2022-11-14 21:24:34,381 INFO mapreduce.JobSubmitter: Executing with tokens: []
2022-11-14 21:24:34,845 INFO mapr

In [23]:
!ls /content/output/shakespeare

part-00000  _SUCCESS


In [24]:
!cat /content/output/shakespeare/part-00000

A saída de streaming foi truncada nas últimas 5000 linhas.
redistribution	1
refund"	1
release	1
replacement	1
resembling	2
reserve	2
return	2
revenge	1
revenue	1
richer	1
rise,	1
robb'd	1
robbing	1
root	1
roses	2
rough	1
royalty	2
ruin	1
s.	4
send	1
shakespeare's	2
sonnets	3
start	1
start:	1
strict	1
such	1
salt	1
sap	1
saturn	1
savage,	1
save	6
save,	1
say	1
section	7
sections	1
see	2
seeking	1
seems	1
self	1
september,	1
service.	1
serving	1
sets	2
shakespeare	4
shakespeare's	3
shall	10
she	3
shifts	1
show	1
showing	1
simply	1
sin	1
since	12
since,	2
sing	1
sings	1
sinks	1
siren	1
sland'ring	1
so	43
so,	4
some	8
sometime	2
sonnets	1
sonnets,	2
speak	2
speaking	1
special	1
spend'st	1
spending	1
states	4
states,	2
states.	4
steal	1
stealing	2
still	2
stirr'd	1
straight	1
strikes	1
such	5
suffering	1
suns	1
supposed	1
suspect	1
swear	1
sweet	5
sweets	1
that	2
the	5
this	5
those	1
to	2
trademark	1
take	3
tan	1
tell	1
tempteth	1
ten	1
terms	2
than	13
that	82
that's	1
that,	1
the	91
thee	1

In [25]:
!wc -l /content/output/shakespeare/part-00000

5662 /content/output/shakespeare/part-00000


Existem, segundo o experimento, 5662 palavras distintas na coleção shakepeare

Gnutella:

In [26]:
!/usr/local/hadoop-3.3.4/bin/hadoop jar /usr/local/hadoop-3.3.4/share/hadoop/tools/lib/hadoop-streaming-3.3.4.jar -input /content/collections/p2p-Gnutella08.txt -output /content/output/gnutella -file /content/mapper-v2.py  -file /content/reducer-v2.py  -mapper 'python mapper-v2.py'  -reducer 'python reducer-v2.py'

2022-11-14 21:24:55,816 WARN streaming.StreamJob: -file option is deprecated, please use generic option -files instead.
packageJobJar: [/content/mapper-v2.py, /content/reducer-v2.py] [] /tmp/streamjob4548729842223011850.jar tmpDir=null
2022-11-14 21:24:56,646 INFO impl.MetricsConfig: Loaded properties from hadoop-metrics2.properties
2022-11-14 21:24:56,867 INFO impl.MetricsSystemImpl: Scheduled Metric snapshot period at 10 second(s).
2022-11-14 21:24:56,867 INFO impl.MetricsSystemImpl: JobTracker metrics system started
2022-11-14 21:24:56,898 WARN impl.MetricsSystemImpl: JobTracker metrics system already initialized!
2022-11-14 21:24:57,077 INFO mapred.FileInputFormat: Total input files to process : 1
2022-11-14 21:24:57,107 INFO mapreduce.JobSubmitter: number of splits:1
2022-11-14 21:24:57,431 INFO mapreduce.JobSubmitter: Submitting tokens for job: job_local2086712167_0001
2022-11-14 21:24:57,431 INFO mapreduce.JobSubmitter: Executing with tokens: []
2022-11-14 21:24:57,878 INFO mapr

In [27]:
!cat /content/output/gnutella/part-00000

#	4
(each	1
0	10
1	1
10	7
100	4
1000	2
1001	20
1002	13
1003	4
1004	12
1005	13
1006	2
1007	4
1008	1
1009	11
101	14
1010	14
1011	2
1012	1
1013	10
1014	1
1015	11
1016	4
1017	6
1018	5
1019	15
102	17
1020	4
1021	11
1022	25
1023	12
1024	3
1025	15
1026	17
1027	16
1028	14
1029	14
103	7
1030	15
1031	3
1032	3
1033	11
1034	10
1035	1
1036	7
1037	4
1038	3
1039	15
104	3
1040	13
1041	12
1042	4
1043	5
1044	8
1045	1
1046	10
1047	10
1048	18
1049	5
105	13
1050	8
1051	2
1052	5
1053	1
1054	8
1055	11
1056	13
1057	11
1058	3
1059	12
106	9
1060	4
1061	11
1062	2
1063	11
1064	9
1065	1
1066	4
1067	11
1068	12
1069	2
107	6
1070	5
1071	3
1072	2
1073	1
1074	1
1075	9
1076	2
1077	5
1078	2
1079	2
108	19
1080	2
1081	12
1082	12
1083	1
1084	20
1085	7
1086	13
1087	10
1088	3
1089	14
109	10
1090	15
1091	11
1092	2
1093	2
1094	14
1095	4
1096	2
1097	7
1098	9
1099	1
11	14
110	18
1100	6
1101	1
1102	2
1103	7
1104	2
1105	2
1106	2
1107	3
1108	9
1109	2
111	14
1110	1
1111	11
1112	1
1113	10
1114	2
1115	8
1116	2
1117	10
1118	12
1119	1
11

In [28]:
!wc -l /content/output/gnutella/part-00000

6324 /content/output/gnutella/part-00000


Existem, segundo o experimento, 6324 palavras distintas na coleção gnutella

### Questão 2
Liste as 10 palavras que mais ocorrem na coleção


In [29]:
import re
rgx = re.compile(r'(.+)\t(\d+)\n')
def getWordMap(collection):
  with open(f'output/{collection}/part-00000', 'r') as f:
    map = {}
    lines = f.readlines()
    for line in lines:
      result = rgx.search(line)
      map[result.group(1)] = int(result.group(2))
    return {k: v for k, v in sorted(map.items(), key=lambda item: item[1], reverse=True)}

def showTop10():
  collections = ['shakespeare','gnutella']
  for collection in collections:
    map = getWordMap(collection)
    top10 = list(map.items())[:10]
    print(f'** Top 10 palavras da coleção {collection} **')
    for word, rank in top10:
      print(f'palavra: {word}     quantidade: {rank}')
    print('-'*50)

showTop10()

** Top 10 palavras da coleção shakespeare **
palavra: the     quantidade: 513
palavra: of     quantidade: 456
palavra: to     quantidade: 401
palavra: my     quantidade: 341
palavra: in     quantidade: 338
palavra: i     quantidade: 327
palavra: and     quantidade: 316
palavra: thy     quantidade: 252
palavra: that     quantidade: 248
palavra: with     quantidade: 209
--------------------------------------------------
** Top 10 palavras da coleção gnutella **
palavra: 123     quantidade: 97
palavra: 127     quantidade: 95
palavra: 367     quantidade: 94
palavra: 424     quantidade: 92
palavra: 251     quantidade: 91
palavra: 264     quantidade: 91
palavra: 266     quantidade: 91
palavra: 427     quantidade: 91
palavra: 145     quantidade: 90
palavra: 249     quantidade: 90
--------------------------------------------------


### Questão 3
Apresente um histograma da frequência de ocorrência das palavras na coleção. O histograma deverá ter 10 classes, cada uma representando um intervalo de frequência. O formato do histograma deverá ser textual, e não gráfico. Por exemplo, seja Ii o í-ésimo intervalo de frequência e Cn número de palavras cuja frequência de ocorrência está no intervalo, o histograma teria o seguinte formato:

I1 - C1  
I2 - C2  
In - Cn  

In [30]:
import numpy as np

def showHistogram():
  collections = ['shakespeare','gnutella']
  for collection in collections:
    map = getWordMap(collection)
    hist, bins = np.histogram(list(map.values()), range = (0.0, 520.0))
    print(f'** Histograma da coleção {collection} **')
    for i in range(len(hist)):
      print(f'l{i+1}[{bins[i]}, {bins[i+1]}]    C{i+1} = {hist[i]}')
    print('-'*50)

showHistogram()

** Histograma da coleção shakespeare **
l1[0.0, 52.0]    C1 = 5219
l2[52.0, 104.0]    C2 = 21
l3[104.0, 156.0]    C3 = 7
l4[156.0, 208.0]    C4 = 4
l5[208.0, 260.0]    C5 = 3
l6[260.0, 312.0]    C6 = 0
l7[312.0, 364.0]    C7 = 4
l8[364.0, 416.0]    C8 = 1
l9[416.0, 468.0]    C9 = 1
l10[468.0, 520.0]    C10 = 1
--------------------------------------------------
** Histograma da coleção gnutella **
l1[0.0, 52.0]    C1 = 6274
l2[52.0, 104.0]    C2 = 50
l3[104.0, 156.0]    C3 = 0
l4[156.0, 208.0]    C4 = 0
l5[208.0, 260.0]    C5 = 0
l6[260.0, 312.0]    C6 = 0
l7[312.0, 364.0]    C7 = 0
l8[364.0, 416.0]    C8 = 0
l9[416.0, 468.0]    C9 = 0
l10[468.0, 520.0]    C10 = 0
--------------------------------------------------


## Modificando a Versão Original
Baixar versão do mapper com regex

In [31]:
!wget https://raw.githubusercontent.com/igroks/bd2/main/projeto-de-impletacao-1/mapper-v2-re.py
!chmod u+rwx /content/mapper-v2-re.py

--2022-11-14 21:26:36--  https://raw.githubusercontent.com/igroks/bd2/main/projeto-de-impletacao-1/mapper-v2-re.py
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 444 [text/plain]
Saving to: ‘mapper-v2-re.py’

mapper-v2-re.py     100%[===================>]     444  --.-KB/s    in 0s      

2022-11-14 21:26:36 (12.9 MB/s) - ‘mapper-v2-re.py’ saved [444/444]



### Questão 4
Responder a Questão 1 com a nova versão

In [32]:
# limpando arquivos de saída
!rm -r output

Shakespeare:

In [33]:
!/usr/local/hadoop-3.3.4/bin/hadoop jar /usr/local/hadoop-3.3.4/share/hadoop/tools/lib/hadoop-streaming-3.3.4.jar -input /content/collections/shakespeare.txt -output /content/output/shakespeare -file /content/mapper-v2-re.py  -file /content/reducer-v2.py  -mapper 'python mapper-v2-re.py'  -reducer 'python reducer-v2.py'

2022-11-14 21:26:46,150 WARN streaming.StreamJob: -file option is deprecated, please use generic option -files instead.
packageJobJar: [/content/mapper-v2-re.py, /content/reducer-v2.py] [] /tmp/streamjob17330490282901289743.jar tmpDir=null
2022-11-14 21:26:47,005 INFO impl.MetricsConfig: Loaded properties from hadoop-metrics2.properties
2022-11-14 21:26:47,261 INFO impl.MetricsSystemImpl: Scheduled Metric snapshot period at 10 second(s).
2022-11-14 21:26:47,261 INFO impl.MetricsSystemImpl: JobTracker metrics system started
2022-11-14 21:26:47,285 WARN impl.MetricsSystemImpl: JobTracker metrics system already initialized!
2022-11-14 21:26:47,435 INFO mapred.FileInputFormat: Total input files to process : 1
2022-11-14 21:26:47,456 INFO mapreduce.JobSubmitter: number of splits:1
2022-11-14 21:26:47,729 INFO mapreduce.JobSubmitter: Submitting tokens for job: job_local1856746628_0001
2022-11-14 21:26:47,729 INFO mapreduce.JobSubmitter: Executing with tokens: []
2022-11-14 21:26:48,173 INFO 

In [34]:
!ls /content/output/shakespeare

part-00000  _SUCCESS


In [35]:
!cat /content/output/shakespeare/part-00000

a	17
actual,	1
agree	2
agreement	1
ak,	1
and	1
any	3
anything	1
ascii"	2
about	1
above	1
accuse	1
additional	1
admit	1
adonis,	1
advantage	1
aeronautical	2
after	2
against	10
ah!	5
al	2
alack!	1
alas!	2
all	10
although	5
am	1
among	1
an	1
and	239
and,	3
angry	1
anon	1
another	1
any	1
applying	1
april	3
april's	1
april,	1
archive	13
are	3
art	2
art's	1
as	38
as,	2
askance	1
at	5
attending	1
author:	1
authorizing	1
awakes	1
ay	1
ay,	1
b.	1
be	1
before	1
breach	2
but	1
bare	1
be	8
be,	1
bear	1
bearing	2
beated	1
beauteous	1
beauty	2
beauty's	1
because	2
before	2
before,	1
beggar'd	1
being	4
beshrew	1
better	1
betwixt	1
beyond	1
blessed	1
book	1
borne	1
both	3
bound	1
breathed	1
bring	1
but	84
but,	5
buy	1
by	15
c	1
ci	1
cii	1
ciii	1
civ	1
cix	1
cl	1
cli	1
clii	1
cliii	1
cliv	1
consequential,	1
contract	1
cv	1
cvi	1
cvii	1
cviii	1
cx	1
cxi	1
cxii	1
cxiii	1
cxiv	1
cxix	1
cxl	1
cxli	1
cxlii	1
cxliii	1
cxliv	1
cxlix	1
cxlv	1
cxlvi	1
cxlvii	1
cxlviii	1
cxv	1
cxvi	1
cxvii	1
cxviii	1
cxx	1
cxxi	

In [36]:
!wc -l /content/output/shakespeare/part-00000

5552 /content/output/shakespeare/part-00000


Existem, segundo o experimento com a nova versão, 5552 palavras distintas na coleção shakespeare

Gnutella:

In [37]:
!/usr/local/hadoop-3.3.4/bin/hadoop jar /usr/local/hadoop-3.3.4/share/hadoop/tools/lib/hadoop-streaming-3.3.4.jar -input /content/collections/p2p-Gnutella08.txt -output /content/output/gnutella -file /content/mapper-v2-re.py  -file /content/reducer-v2.py  -mapper 'python mapper-v2-re.py'  -reducer 'python reducer-v2.py'

2022-11-14 21:27:10,749 WARN streaming.StreamJob: -file option is deprecated, please use generic option -files instead.
packageJobJar: [/content/mapper-v2-re.py, /content/reducer-v2.py] [] /tmp/streamjob17361670238180491009.jar tmpDir=null
2022-11-14 21:27:11,554 INFO impl.MetricsConfig: Loaded properties from hadoop-metrics2.properties
2022-11-14 21:27:11,787 INFO impl.MetricsSystemImpl: Scheduled Metric snapshot period at 10 second(s).
2022-11-14 21:27:11,787 INFO impl.MetricsSystemImpl: JobTracker metrics system started
2022-11-14 21:27:11,815 WARN impl.MetricsSystemImpl: JobTracker metrics system already initialized!
2022-11-14 21:27:11,999 INFO mapred.FileInputFormat: Total input files to process : 1
2022-11-14 21:27:12,025 INFO mapreduce.JobSubmitter: number of splits:1
2022-11-14 21:27:12,361 INFO mapreduce.JobSubmitter: Submitting tokens for job: job_local99742931_0001
2022-11-14 21:27:12,362 INFO mapreduce.JobSubmitter: Executing with tokens: []
2022-11-14 21:27:12,779 INFO ma

In [38]:
!cat /content/output/gnutella/part-00000

august	1
directed	2
edges:	1
fromnodeid	1
gnutella	1
nodes:	1
p2p	1
tonodeid	1
from	1
graph	1
is	1
network	1
nodes	1
of	1
once):	1
p2p-gnutella08.txt	1
pair	1
saved	1
unordered	1


In [39]:
!wc -l /content/output/gnutella/part-00000

19 /content/output/gnutella/part-00000


Existem, segundo o experimento com a nova versão, 19 palavras distintas na coleção gnutella

### Questão 5
Responder a Questão 2 com a nova versão

In [40]:
showTop10()

** Top 10 palavras da coleção shakespeare **
palavra: the     quantidade: 513
palavra: of     quantidade: 456
palavra: to     quantidade: 401
palavra: my     quantidade: 341
palavra: in     quantidade: 338
palavra: i     quantidade: 327
palavra: and     quantidade: 316
palavra: thy     quantidade: 252
palavra: that     quantidade: 248
palavra: with     quantidade: 209
--------------------------------------------------
** Top 10 palavras da coleção gnutella **
palavra: directed     quantidade: 2
palavra: august     quantidade: 1
palavra: edges:     quantidade: 1
palavra: fromnodeid     quantidade: 1
palavra: gnutella     quantidade: 1
palavra: nodes:     quantidade: 1
palavra: p2p     quantidade: 1
palavra: tonodeid     quantidade: 1
palavra: from     quantidade: 1
palavra: graph     quantidade: 1
--------------------------------------------------


### Questão 6
Responder a Questão 3 com a nova versão

In [41]:
showHistogram()

** Histograma da coleção shakespeare **
l1[0.0, 52.0]    C1 = 5111
l2[52.0, 104.0]    C2 = 21
l3[104.0, 156.0]    C3 = 7
l4[156.0, 208.0]    C4 = 4
l5[208.0, 260.0]    C5 = 3
l6[260.0, 312.0]    C6 = 0
l7[312.0, 364.0]    C7 = 4
l8[364.0, 416.0]    C8 = 1
l9[416.0, 468.0]    C9 = 1
l10[468.0, 520.0]    C10 = 1
--------------------------------------------------
** Histograma da coleção gnutella **
l1[0.0, 52.0]    C1 = 19
l2[52.0, 104.0]    C2 = 0
l3[104.0, 156.0]    C3 = 0
l4[156.0, 208.0]    C4 = 0
l5[208.0, 260.0]    C5 = 0
l6[260.0, 312.0]    C6 = 0
l7[312.0, 364.0]    C7 = 0
l8[364.0, 416.0]    C8 = 0
l9[416.0, 468.0]    C9 = 0
l10[468.0, 520.0]    C10 = 0
--------------------------------------------------
